internal/cmd/gocdk/launch.go

```diff
@@ -15,11 +15,170 @@
-	return xerrors.New("not implemented")
+	f := newFlagSet(pctx, "launch")
+	imageName := f.String("image", "", "Docker image name to launch (defaults to project latest)")
+	if err := f.Parse(args); xerrors.Is(err, flag.ErrHelp) {
+		return nil
+	} else if err != nil {
+		return usagef("gocdk launch: %w", err)
+	}
+	if f.NArg() != 1 {
+		return usagef("gocdk launch BIOME")
+	}
+	biome := f.Arg(0)
+
+	moduleRoot, err := findModuleRoot(ctx, pctx.workdir)
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+
+	// Get the image name from the Dockerfile if not specified.
+	if *imageName == "" {
+		var err error
+		*imageName, err = moduleDockerImageName(moduleRoot)
+		if err != nil {
+			return xerrors.Errorf("gocdk launch: %w", err)
+		}
+	}
+
+	// Prepare the launcher.
+	cfg, err := readBiomeConfig(moduleRoot, biome)
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+	if cfg.Launcher == nil {
+		return xerrors.Errorf("gocdk launch: launcher not specified in %s", filepath.Join(findBiomeDir(moduleRoot, biome), biomeConfigFileName))
+	}
+	launcher, err := newLauncher(ctx, pctx, *cfg.Launcher)
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+
+	// Read the launch specifier from the biome's Terraform output.
+	tfOutput, err := tfReadOutput(ctx,
+		findBiomeDir(moduleRoot, biome),
+		pctx.overrideEnv("TF_IN_AUTOMATION=1"))
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+
+	// Launch the application.
+	launchURL, err := launcher.Launch(ctx, &launchInput{
+		dockerImage: *imageName,
+		specifier:   launchSpecifier(tfOutput["launch_specifier"].mapValue()),
+		// TODO(light): Pass environment variables from tfOutput into env.
+	})
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+	fmt.Fprintf(pctx.stdout, "Serving at %s\n", launchURL)
+	return nil
+}
+
+// Launcher is the interface for any type that can launch a Docker image.
+type Launcher interface {
+	Launch(ctx context.Context, input *launchInput) (*url.URL, error)
+}
+
+// launchInput is the input to a launcher.
+type launchInput struct {
+	// dockerImage specifies the image name and tag of the local Docker image to
+	// deploy. If the local image does not exist, then the launcher should return
+	// an error.
+	dockerImage string
+
+	// env is the set of additional environment variables to set. It should not
+	// include PORT nor should it contain multiple entries for the same variable
+	// name.
+	env []string
+
+	// specifier is the set of arguments passed from a biome's Terraform module.
+	specifier launchSpecifier
+}
+
+// newLauncher creates the launcher for the given name.
+func newLauncher(ctx context.Context, pctx *processContext, launcherName string) (Launcher, error) {
+	logger := log.New(pctx.stderr, "gocdk: ", log.Ldate|log.Ltime)
+	switch launcherName {
+	case "local":
+		return &localLauncher{
+			logger:    logger,
+			dockerEnv: pctx.env,
```

Comment:
> It's confusing to me that `localLauncher` has a `dockerEnv` from `pctx.env`, but `launchInput` also has an `env`. What goes where? Can we just have one set of environment variables?

Based on discussions around #1890, using environment variables to pass configuration to the running binary seems right, so these are two distinct sets of environment variables. We can't join them.

> IIUC, one is the set used to exec `docker`, while the other is the set passed to the docker image being run. I guess that's reasonable. Maybe `localLauncher` can just have a `processContext` field instead of copying its subfields?

I was trying to be as specific as possible with what the `localLauncher` depends on from the `processContext`.

Reply:
I guess so, but it makes it less readable since you're basically aliasing a bunch of things -- I have to keep track of the fact that `dockerEnv` is really `pctx.env`. There are so many different envs and dirs floating around that extra names for existing ones just adds to the confusion.

Reply:
Ack. I'd like to address this as part of #2046.